Investigate noIterationTypes crash

rbuckton opened the pull request:
Investigating a crash related to the `noIterationTypes` sentinel.

Fixes #32953

rbuckton commented:
@typescript-bot pack this

rbuckton commented:
@typescript-bot pack this

rbuckton commented:
@typescript-bot pack this

rbuckton commented:
@typescript-bot pack this

rbuckton commented:
@dbaeumer, @falsandtru: Once @typescript-bot indicates the tarball for the proposed fix has completed, would you be able to try it out in your local workspace and report back as to whether this has improved stability in the editor?

falsandtru commented:
Looks like the problem disappeared :+1:

rbuckton commented:
I'd like to verify whether this issue was indeed related to `nameType` being incorrectly set on a `Symbol`. After the bot finishes packing up the new changes, I'd appreciate if you were able to try the new tarball to see if the issue is still resolved.

@typescript-bot pack this

falsandtru commented:
I can repro with the reverted version. As I wrote in #32953, I can repro also with yesterday's build.

rbuckton commented:
Unfortunately, I still am unable to repro this issue following the steps you provided.

rbuckton commented:
@typescript-bot pack this

rbuckton commented:
@falsandtru, @dbaeumer: Unfortunately I've been unable to repro this using either @falsandtru's repro steps or by manually making changes to the ESLint VSCode extension that @dbaeumer was working on in their issue.

While it looks like we have [verified](https://github.com/microsoft/TypeScript/pull/36388#issuecomment-577965942) that we can resolve this specific issue by moving `noIterationTypes` outside of `createTypeChecker`, the fact that we could end up with multiple *different* versions of `noIterationTypes` indicates there is a leak somewhere in the type checker, as this should not be possible.

So far I have identified 3 possible culprits that could leak a `TypeChecker` (and an older `noIterationTypes`):

- The `nameType` property on `Symbol`, which could leak a type on symbols created by the binder that could be reused during incremental parse.
  - We have already verified that this was [not the underlying cause](https://github.com/microsoft/TypeScript/pull/36388#issuecomment-578026816) in @falsandtru's case.
- The `contextualType` property on `Node` could outlive a `TypeChecker` if the current semantic operation is canceled (via a cancellation token).
- The `combineValueAndTypeSymbols` function could possibly assign a `SymbolTable` created by the binder to a `TransientSymbol` created by the checker. This could also result in a leak as much of the checker assumes that symbol tables on transient symbols are themselves transient and won't outlive the checker.

I've just pushed up a commit that adds a number of checks to the `TypeChecker` public API and several internal APIs, as well as addressed the issues mentioned above. The hope is that either one of the above issues was the culprit, or that we will be able to use the additional checks to narrow down the underlying leak. If you are able, once the bot has finished packing a tarball of these changes, I would appreciate if you could use it to test your respective scenarios.

falsandtru commented:
Not reproducible as expected.

rbuckton commented:
The issue no longer exists as of this build? That indicates the issue was one of the three mentioned above.

falsandtru commented:
Yes. The build for debugging should output debug log to discover the problem.

falsandtru commented:
Another, the log of https://github.com/microsoft/TypeScript/issues/32953#issuecomment-577958476 doesn't help to investigate how to repro?

dbaeumer commented:
@rbuckton we have endgame this week so not too much programming. But I will make sure I use the latest build and have an eye on stale problems. From my experience I think I got hit by the cancellation. The traces even showed that a semantic diagnostic operation got canceled before. So may it can be reproduce by setting a break point into the server that results in the client to cancel the operation.

rbuckton commented:
Unfortunately I don't have a test that can repro this behavior, so I am hoping at the very least that the guards that I've added will be enough to catch future leaks of this kind.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -317,21 +317,6 @@ namespace ts {
         const emitResolver = createResolver();
         const nodeBuilder = createNodeBuilder();
 
-        const globals = createSymbolTable();
```
NOTE: I needed to move these after the `const checker = {...}` declaration as `createSymbol` now depends on `checker` to ensure we add the current `TypeChecker` to all symbols to make sure we are always using the correct checker.

rbuckton commented:
@typescript-bot perf test

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -351,10 +336,14 @@ namespace ts {
             isUnknownSymbol: symbol => symbol === unknownSymbol,
-            getMergedSymbol,
+            getMergedSymbol: symbol => {
+                verifyTypeCheckerOfSymbol(symbol);
```
How about having `verifyCheckerOfXXX` return the value it is passed such that you can write `return getMergedSymbol(verifyCheckerOfSymbol(symbol))` and keep it to just one expression.

ahejlsberg commented on src/compiler/utilities.ts:
```diff
@@ -4972,20 +4972,17 @@ namespace ts {
         this.id = undefined;
         this.mergeId = undefined;
         this.parent = undefined;
+        this.checker = undefined;
```
Not crazy about this *always* being there.

ahejlsberg commented:
I'm not totally sold on the value of the `verifyCheckerOfXXX` calls and the `checker` property always being present. It's overhead and the bug we're fixing wasn't actually caused by anything this is verifying.

rbuckton commented:
I can pull the API verification for now, but we really should consider ensuring our public API doesn't allow you to try to mix and match types/signatures/symbols produces by other checkers.

rbuckton commented:
@ahejlsberg: Any other comments after having reverted the `verifyX` changes for now?

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -2277,8 +2278,8 @@ namespace ts {
-            if (typeSymbol.members) result.members = typeSymbol.members;
-            if (valueSymbol.exports) result.exports = valueSymbol.exports;
+            if (typeSymbol.members) result.members = cloneMap(typeSymbol.members);
+            if (valueSymbol.exports) result.exports = cloneMap(valueSymbol.exports);
```
Why do we need to clone here?

weswigham replied:
if `typeSymbol` or `valueSymbol` are _not_ `TransietSymbol`s, then their symbol tables consist only of non-transient symbols and persist through multiple checks. If we only copy the reference to the table into a new `TransientSymbol`, when we later stuff more things into the table (as is possible for transient symbols), we'll pollute the supposedly "immutable" symbol table from the original non-transient symbol with new transient members.

rbuckton replied:
This was one of 3 possible leaks. We assume elsewhere in the checker that any `SymbolTable` attached to a `TransientSymbol` is safe to mutate, and sometimes add new symbols to it. `combineValueAndTypeSymbols` currently just blindly assigns the reference to the symbol table to `result` (which is a `TransientSymbol`), ignoring whether `typeSymbol` or `valueSymbol` were created by the binder. As a result, there is the possibility of some other function in checker later mutating the `SymbolTable`.

rbuckton commented:
Need to reopen this on master since I used release-3.8 primarily for providing a stable drop for testers.

@typescript-bot cherry-pick this to master

rbuckton commented:
Closing in favor of https://github.com/microsoft/TypeScript/pull/36491, which ports this change to `master` instead.

rbuckton commented:
This has been merged into `master`.
